### PR TITLE
[WFLY-16658] Update some obsolete contents in wildfly extension documentation

### PR DESCRIPTION
**Pull request by josevjunior**
This commit change the extension archetype identification and update the demo project that are present in extending wildfly's documentation

Jira issue: https://issues.redhat.com/browse/WFLY-16658

### CODE REVIEW

**Comment by yersan** [comment]
Hello @josevjunior  -- thanks for the submission!

We require that all contributions be made under the terms of the MIT License, http://www.opensource.org/licenses/mit-license.php. Do you agree to these terms?

**Comment by josevjunior** [comment]
Hello, @yersan 
I do agree.

**Review by yersan — CHANGES_REQUESTED** [review]
Hi @josevjunior, thank you for looking at this document. I've added some comments reviewing the changes. In addition, I've found other details in this documentation that is affected by using the new archetype. See below, I've added the Lines number that corresponds to your `Example_subsystem.adoc` version:

----

Example_subsystem.adoc:668
Fix wording `operationa`
Remove "But we need to implement the `DescriptionProvider` method to provide the model description". The description provider is automatically added for the remove handler (similar to the Add Handler) so we don't need to implement a `DescriptionProvider`.

----

Example_subsystem.adoc:808 - Example_subsystem.adoc:816

"We then add a new step associated with the `RUNTIME` stage to update the
tick of the TrackerService for our suffix. This is essential since the
call to `context.getServiceRegistry()` will fail unless the step
accessing it belongs to the `RUNTIME` stage."

We can remove this paragraph and the following note since we do not implement the `execute()` method. This is done by the `AbstractWriteAttributeHandler` class.

----

We should update the pieces of code for the test cases that are using the following code with the one generated by the new archetype:

`KernelServices services = super.installInController(subsystemXml);`

For example:
Example_subsystem.adoc:1124
Example_subsystem.adoc:1167
Example_subsystem.adoc:1180
Example_subsystem.adoc:1269
Example_subsystem.adoc:1319

----

Example_subsystem.adoc:1195 - xample_subsystem.adoc:1243

These lines can be removed since `testDescribeHandler()` is no longer generated by the archetype, we are not testing the describe operation.

----

Example_subsystem.adoc:1413

Replace `SubsystemAddHandler` with `SubsystemAdd`, by default `SubsystemAdd` is generated by the archetype.

----

Example_subsystem.adoc:1708

"will return a lot of information, including what we provided in the `DescriptionProvider`s we created to document our subsystem."

We are no longer adding manually a `DescriptionProvider` since it is automatically configured for us, so I think we can just say the following:

"will return a lot of information, including a description of our subsystem generated automatically by a `DescriptionProvider` " 

----

You have added a new zip ([acme-subsystem_v1.zip](https://github.com/wildfly/wildfly/pull/15813/files#diff-12357c2dd4ecf3cb1021f473d57d6f4dbaa11659b4983deef36d661d893c38e1)) you have to use the original name and replace the existing one.

I've checked your zip file and the `pom.xml` you are providing differs a lot from the one generated by the archetype.

I've just followed the tutorial from scratch and I did not need to change the `pom.xml` generated by default. I would say you can use the one generated by the archetype.

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 1]
```diff
@@ -39,11 +39,11 @@ $ mvn archetype:generate \
  
 .........
  
-Define value for property 'groupId': : com.acme.corp
-Define value for property 'artifactId': : acme-subsystem
+Define value for property 'module': com.acme.corp.tracker
+Define value for property 'groupId': com.acme.corp
+Define value for property 'artifactId': acme-subsystem
```
[the remark this answers is not in the document]
Comment (josevjunior) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc:44]: Really? I'm using that mvn version too and here the module property is the first do appear. That's why I changed the order
image

My env:
image

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 2]
```diff
@@ -615,19 +607,19 @@ the deployments.
...
+}
  
-    @Override
-    public void stop(StopContext context) {
-        OUTPUT.interrupt();
-    }
+@Override
+public void stop(StopContext context) {
+    OUTPUT.interrupt();
+}
```
[the remark this answers is not in the document]
Comment (josevjunior) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc:622]: Can you explain a little bit?

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 3]
```diff
@@ -305,26 +322,8 @@ The `ManagementResourceRegistration` obtained in
...
-3.  The handler description provider - we will talk more about this
-below.
-4.  Whether this operation handler is inherited - `false` means that
-this operation is not inherited, and will only apply to
-`/subsystem=tracker`. The content for this operation handler will be
-provided by `3`.
+achieved by providing it in constructor of your `ResourceDefinition` 
+just as we did in example above.
 
 Let us first look at the description provider which is quite simple
```
[the remark this answers is not in the document]
Comment (josevjunior) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc:328]: I'll read the text again and try to adapt it

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 4]
```diff
@@ -227,6 +227,20 @@ public class SubsystemExtension implements Extension {
     ...
 ----
 
+Modify the src/main/resources/com/acme/corp/tracker/LocalDescriptions.properties
```
[the remark this answers is not in the document]
Comment (josevjunior) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc:230]: Ok. Will be done in the next commit

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc:44]: I'm running it on Linux, so maybe that makes the difference. Don't worry about it. It's ok if it runs in a different order. We can keep your changes, thanks!

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc:622]: Yes. The JBoss Modular Service Container (MSC) API was improved some time ago to allow multiple values per service. The old API was deprecated and right now we have in the server a mix of services running on legacy and new API.

The service used in this tutorial is based on the legacy MSC API; it extends from `org.jboss.msc.service.Service`, which [is deprecated](https://github.com/jboss-msc/jboss-msc/blob/master/src/main/java/org/jboss/msc/service/Service.java#L47)

It works without any issues, indeed, the server internally is still using tons of these services, but for new features and developments we have to extend from `org.jboss.msc.Service` instead which uses a different programming model. This is for sure out of the task you are doing here, I'll open a Jira to address it. Don't worry about this on this PR.

If you want to learn more, the MSC tessuite contains useful pieces of code that could help to understand the differences between the legacy and the new programming style. The [MultiValueServicesTestCase](https://github.com/jboss-msc/jboss-msc/blob/master/src/test/java/org/jboss/msc/multi_value_services/MultiValueServicesTestCase.java#L96) test contains a mix of both.

**Comment by yersan** [comment]
> @yersan The new zip package I uploaded got a little bit confusing. The acme-subsystem_v1 is the renamed old one and the new one remains with the same name. I did not specify this in the commit.

Ahh, got it, thanks I did not realize about it. In that case, you can completely discard the old zip, we don't need to backup it as a separate artifact.

**Comment by josevjunior** [comment]
> > @yersan The new zip package I uploaded got a little bit confusing. The acme-subsystem_v1 is the renamed old one and the new one remains with the same name. I did not specify this in the commit.
> 
> Ahh, got it, thanks I did not realize about it. In that case, you can completely discard the old zip, we don't need to backup it as a separate artifact.

Can we turn the acme-subsystem source to be git managed?

**Comment by yersan** [comment]
> Can we turn the acme-subsystem source to be git managed?

Having a zip file facilitates downloading it when you are following the tutorial. This application is just an example that is part of the tutorial and it should not change too often, only when the tutorial changes.

For that purpose, we would need a separate GitHub repo to store it (and maintain it) and that makes things more difficult to track (tutorial doc and example application separated in different repos).

Otherwise, maybe we could have a specific directory inside the WildFly docs module for having the source code of this application. This directory must be completely detached from the docs or any WildFly build system, but in that case, users would need to clone the WildFly project to follow the tutorial. 

None of the above makes me think of a good benefit for doing that.

@josevjunior What's your motivation for doing it?

**Comment by josevjunior** [comment]
> @josevjunior What's your motivation for doing it?

My idea was get easier to track the changes in the example project. But thinking a little bit, this would not prevent to mistakes be made, once the zip should've be pushed manually after each modification.

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (josevjunior) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc]: Changes was pushed in the new commit made yesterday

**Comment by josevjunior** [comment]
Hi, @yersan . Some time ago you ask me to fit my changes to the [repository style guide](https://docs.wildfly.org/26/Hacking_On_WildFly.html#ide-integration) Can you be a little more specific about what I must change? Is it about the acme-subsytem template?

**Comment by yersan** [comment]
Hi @josevjunior, yes sure, we need that the source code of the classes you are including on [acme-subsystem.zip](https://github.com/wildfly/wildfly/pull/15813/files#diff-a4568a272193ca7b34ef2a2863e9c12e15146192233b84597ab02a3be9ae6fce) be formatted with the standard code style we use in WildFly. This code style is based on Eclipse IDE code style rules.

The [repository style guide](https://docs.wildfly.org/26/Hacking_On_WildFly.html#ide-integration) describes how you can import this code style formatter on Eclipse. Once you have your IDE configured accordingly, apply the format to all the files you are including in the acme-subsystem.zip. 

If you are using any other IDE, check if there is any plugin that allows you to import Eclipse IDE code style rules into it.

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 5]
```diff
@@ -1415,7 +1369,7 @@ to install the deployers, which is done in the following method:
         //see SubDeploymentProcessor for explanation of the phases
         context.addStep(new AbstractDeploymentChainStep() {
             public void execute(DeploymentProcessorTarget processorTarget) {
-                processorTarget.addDeploymentProcessor(SubsystemDeploymentProcessor.PHASE, SubsystemDeploymentProcessor.priority, new SubsystemDeploymentProcessor());
+                processorTarget.addDeploymentProcessor(SubsystemExtension.SUBSYSTEM_NAME, SubsystemDeploymentProcessor.PHASE, SubsystemDeploymentProcessor.priority, new SubsystemDeploymentProcessor());
```
Comment (yersan) [inline docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc:1372]: Could you replace `SubsystemDeploymentProcessor.priority` with `SubsystemDeploymentProcessor.PRIORITY`?

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc (hunk 3 above)
[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc]: Resolved

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc (hunk 4 above)
[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc]: Resolved

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 6]
```diff
@@ -415,14 +401,14 @@ operation handlers is that we do any model updates before changing the
...
-removed so we have a remove handler for the subsystem registered +
+removed. So we have a remove handler for the subsystem registered +
 in `SubsystemDefinition.registerOperations` or just provide the
 operation handler in constructor.
 
 [source,java,options="nowrap"]
 ----
 //Every resource that is added, normally needs a remove operation
-        registration.registerOperationHandler(REMOVE, SubsystemRemove.INSTANCE, DefaultResourceRemoveDescriptionProvider(resourceRegistration,descriptionResolver) , false);
+resourceRegistration.registerOperationHandler(new SimpleOperationDefinitionBuilder(ModelDescriptionConstants.REMOVE, SubsystemExtension.getResourceDescriptionResolver(null)).build(), SubsystemAdd.INSTANCE);
```
[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc]: Resolved

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 7]
```diff
@@ -540,21 +531,22 @@ updated before runtime changes are made.
...
 [source,java,options="nowrap"]
 ----
-@Override
-    protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model,
-            ServiceVerificationHandler verificationHandler, List<ServiceController<?>> newControllers)
-            throws OperationFailedException {
+public class TypeAddHandler extends AbstractAddStepHandler {
+    ...
+    @Override
+    protected void performBoottime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
```
[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc]: Resolved

File: docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc [hunk 8]
```diff
@@ -698,18 +690,38 @@ public class TypeRemoveHandler extends AbstractRemoveStepHandler {
 }
 ----
 
-We then need a description provider for the `type` part of the model
-itself, so we modify TypeDefinitnion to registerAttribute
+We then need a register all the operation for the `type` part of the model
+itself, so we modify TypeDefinitnion to extend SimpleResourceDefinition
```
[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc]: resolved

[the remark this answers is not in the document]
Comment (yersan) [reply docs/src/main/asciidoc/_extending-wildfly/Example_subsystem.adoc]: resolved

**Review by yersan — CHANGES_REQUESTED** [review]
Hi @josevjunior,

Thanks for all the work so far and for taking care of this advanced tutorial. We are pretty close to approving this, I did another round and found the following, in addition to "priority" attributed mentioned in the review:

1. "We then add a new step associated with the RUNTIME stage to update the

tick of the TrackerService for our suffix. This is essential since the
call to context.getServiceRegistry() will fail unless the step
accessing it belongs to the RUNTIME stage."

We can remove this paragraph and the following note since we do not implement the execute() method. This is done by the AbstractWriteAttributeHandler class.

2. Your in your `TypeDefinition`, the TICK attribute is declared as required but also with a default value, change it to `.setRequired(false)` so the default value can be in place

Once these very minor details are fixed I'll give it my approval

**Review by yersan — APPROVED** [review]

**Comment by bstansberry** [comment]
Hi @josevjunior -- thanks for your work on this!

Looking at https://issues.redhat.com/browse/WFLY-16658 I assume it's you who filed it, but your account is shown as 'Inactive'. TBH I don't know what that means. Are you able to log into your account?

I'm asking because I hope to get this merged before we release WF 27, and when we do I'd like to assign the issue to you so the JIRA history will properly reflect that you did the work!

**Pull request merged**